node/impl/full/trace.go

```diff
@@ -0,0 +1,515 @@
+)
+
+type EthTraceAPI interface {
+	TraceBlock(ctx context.Context, blkNum string) (interface{}, error)
+	TraceReplayBlockTransactions(ctx context.Context, blkNum string, traceTypes []string) (interface{}, error)
+}
+
+var (
+	_ EthTraceAPI = *new(api.FullNode)
+)
+
+type EthTrace struct {
+	fx.In
+
+	Chain        *store.ChainStore
+	StateManager *stmgr.StateManager
+
+	ChainAPI
+	EthModuleAPI
+}
+
+var _ EthTraceAPI = (*EthTrace)(nil)
+
+type Trace struct {
+	Action       Action `json:"action"`
+	Result       Result `json:"result"`
+	Subtraces    int    `json:"subtraces"`
+	TraceAddress []int  `json:"traceAddress"`
+	Type         string `json:"Type"`
+
+	parent *Trace
+}
+
+func (t *Trace) setCallType(callType string) {
+	t.Action.CallType = callType
+	t.Type = callType
+}
+
+type TraceBlock struct {
+	*Trace
+	BlockHash           ethtypes.EthHash `json:"blockHash"`
+	BlockNumber         int64            `json:"blockNumber"`
+	TransactionHash     ethtypes.EthHash `json:"transactionHash"`
+	TransactionPosition int              `json:"transactionPosition"`
+}
+
+type TraceReplayBlockTransaction struct {
+	Output          string           `json:"output"`
+	StateDiff       *string          `json:"stateDiff"`
+	Trace           []*Trace         `json:"trace"`
+	TransactionHash ethtypes.EthHash `json:"transactionHash"`
+	VmTrace         *string          `json:"vmTrace"`
+}
+
+type Action struct {
+	CallType string             `json:"callType"`
+	From     string             `json:"from"`
+	To       string             `json:"to"`
+	Gas      ethtypes.EthUint64 `json:"gas"`
+	Input    string             `json:"input"`
+	Value    ethtypes.EthBigInt `json:"value"`
+
+	method  abi.MethodNum `json:"-"`
+	codeCid cid.Cid       `json:"-"`
+}
+
+type Result struct {
+	GasUsed ethtypes.EthUint64 `json:"gasUsed"`
+	Output  string             `json:"output"`
+}
+
+func (e *EthTrace) TraceBlock(ctx context.Context, blkNum string) (interface{}, error) {
+	ts, err := e.getTipsetByBlockNr(ctx, blkNum, false)
+	if err != nil {
+		return nil, err
+	}
+
+	_, trace, err := e.StateManager.ExecutionTrace(ctx, ts)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to compute base state: %w", err)
+	}
+
+	tsParent, err := e.ChainAPI.ChainGetTipSetByHeight(ctx, ts.Height()+1, e.Chain.GetHeaviestTipSet().Key())
+	if err != nil {
+		return nil, fmt.Errorf("cannot get tipset at height: %v", ts.Height()+1)
+	}
+
+	msgs, err := e.ChainGetParentMessages(ctx, tsParent.Blocks()[0].Cid())
+	if err != nil {
+		return nil, err
+	}
+
+	cid, err := ts.Key().Cid()
+	if err != nil {
+		return nil, err
+	}
+
+	blkHash, err := ethtypes.EthHashFromCid(cid)
+	if err != nil {
+		return nil, err
+	}
+
+	allTraces := make([]*TraceBlock, 0, len(trace))
+	for _, ir := range trace {
+		// ignore messages from f00
+		if ir.Msg.From.String() == "f00" {
+			continue
+		}
+
+		idx := -1
+		for msgIdx, msg := range msgs {
+			if ir.Msg.From == msg.Message.From {
```

Comment:
We can have multiple messages from the same sender. We need to instead use the message CIDs. I'd:

1. Build a map of message CID to index.
2. Do the lookup here.

I.e., `ir.MsgCid`, `msg.Cid`.

Reply:
Actually, I don't _think_ this is even necessary? The index in the trace should match the index in the message list as we execute (trace) messages in-order.